src/tdigest.c

```diff
@@ -0,0 +1,512 @@
+	// 'merge' centroid
+	c = &args->centroids[args->index];
+	c->weight += merge->weight;
+	c->mean += (merge->mean - c->mean) * merge->weight / c->weight;
+
+	// update the properties of the being-built tdigest
+	if (merge->weight > 0)
+	{
+		args->min = Min(merge->mean, args->min);
+		args->max = Max(merge->mean, args->max);
+	}
+}
+
+/*
+ * Merges the input buffer of centroids into the t-digest's stored centroids
+ * Implements the merge algorithm, which is algorithm 1 of tdunning's paper
+ */
+TDigest *
+ts_tdigest_compress(TDigest *t, List *unmerged_centroids_list)
+{
+	int num_unmerged = list_length(unmerged_centroids_list);
+	Centroid *unmerged_centroids;
+	uint64_t unmerged_weight = 0;
+	ListCell *lc;
+	int i, j;
+	mergeArgs *args;
+	uint32_t num_centroids = t->num_centroids;
+
+	// return unchanged if no unmerged centroids to add
+	if (!num_unmerged)
+		return t;
+
+	unmerged_centroids = palloc(sizeof(Centroid) * num_unmerged);
+
+	// Move unmerged centroids from linked list to array
+	i = 0;
+	foreach (lc, unmerged_centroids_list)
+	{
+		Centroid *c = (Centroid *) lfirst(lc);
+		memcpy(&unmerged_centroids[i], c, sizeof(Centroid));
+		unmerged_weight += c->weight;
+		i++;
+	}
+
+	list_free_deep(unmerged_centroids_list);
+
+	// return unchanged if centroids are 0-weight
+	if (unmerged_weight == 0)
+		return t;
+
+	t->total_weight += unmerged_weight;
+
+	/* only support up to 2^32 - 1 data points in a TDigest */
+	if (t->total_weight < 0)
+		elog(ERROR, "Too many rows added to TDigest.");
+
+	qsort(unmerged_centroids, num_unmerged, sizeof(Centroid), centroid_cmp);
+
+	// args will store the result of our merge until ready to copy to main t-digest
+	args = palloc0(sizeof(mergeArgs));
+	args->centroids = palloc0(sizeof(Centroid) * t->size);
+	args->t = t;
+	args->min = INFINITY;
+
+	i = 0;
+	j = 0;
+	while (i < num_unmerged && j < t->num_centroids)
+	{
+		Centroid *a = &unmerged_centroids[i];
+		Centroid *b = &t->centroids[j];
+
+		if (a->mean <= b->mean)
+		{
+			merge_centroid(args, a);
+			i++;
+		}
+		else
+		{
+			merge_centroid(args, b);
+			j++;
+		}
+	}
+
+	/*
+	 * Finish running merge on any remaining centroids
+	 */
+	while (i < num_unmerged)
+		merge_centroid(args, &unmerged_centroids[i++]);
+
+	pfree(unmerged_centroids);
+
+	/*
+	 * Finish running merge on remaining centroids from the original t-digest
+	 */
+	while (j < t->num_centroids)
+		merge_centroid(args, &t->centroids[j++]);
+
+	// update properties of the t-digest
+	if (t->total_weight > 0)
+	{
+		t->min = Min(t->min, args->min);
+		t->max = Max(t->max, args->max);
+
+		if (args->centroids[args->index].weight <= 0)
+			args->index--;
+
+		t->num_centroids = args->index + 1;
+	}
+
+	/*
+	 * Adjust memory allocation for t-digest if we have more centroids than we started with
+	 * Necessary to ensure integrity of variable-size array hack used for centroid array in t-digest
+	 */
+	if (t->num_centroids > num_centroids)
```

Comment:
Why are we repalloc-ing here? the alloc is for "size" * Centroid when we create the struct. (Not clear to me what the comment is trying to say either).

Reply:
should rename "size" as max_centroids or something more descriptive. (I mistook size for the size of the alloced struct which it is not).